### DISCUSSION TITLE
Is it a good idea to create a "Snippet Manager" plugin?

### DISCUSSION THREAD

**Discussion by ziontee113** [L3MON4D3/LuaSnip#481] [Ideas]
Hi @L3MON4D3 , thank you so much for the amazing plugin :)

Today I'd like to have a discussion about the snippet creation process.

LuaSnip and the Snippets created from it is awesome, but I can't really say the same for the process of creating those snippets.

Creating a Snippet requires too much context switching & fumbling around to get things done. My "quickest" way is to use `LuaSnipEdit` to jump to the snippet file, press `<CR>` to create some white spaces, then press `jcs` to `create snippet` (trigger a snippet to create snippet), then fumble around thinking of how to name the snippet, it's description, jump around to yank it's contents, more jumping for customizing the nodes using `fmt`, etc... You get the idea.

I'm proposing & planning to create a "Snippet Creator / Manager" in the close future. And I'd love your input on this.
The results it outputs should be just a regular lua snippet file that is 100% compatible with LuaSnip.
The user, in their working buffer, select some text, press a key, the selected text would be replaced with extmarks that looks something like `{xxx}`, informing the user that part will be like the `{}` in the `fmt` snippet creation process.
After marking those `placeholder` regions, the user would run a Command such as `CreateSnippet...`, a pop up window should show up to give the user quick access to the settings they'd be interested in.
I plan to start simple, just supporting insert nodes, simple choice nodes (text nodes / insert nodes as child only), etc... in the beginning.

Please let me know what you think about this idea. Thank you :)

**Comment by L3MON4D3** [comment]
Hey!
Thats an interesting idea! It's true that creating snippets is still pretty complicated, I wholly support anything that makes this process easier 👍
Working with the text that's already in the buffer is a good idea I think, the text-parts at least will be correct already.
Using visualmode+extmarks sounds very cool, I think that's a great idea.
About the popup-window, maybe hydra could be used here? Eg. have keybinds for all kinds of nodes, the marked region the cursor is on if any of those is pressed will be turned into the node.
Another thing that might be cool, but pretty much cool only, would be
1. Showing some representation of the created nodes in Floating windows (actually, maybe this isn't cool only, but pretty useful as one can better keep an overview this way) 
2. Connect dynamic/FunctionNodes with associated insertNodes (not 100% sure how that can be accomplished and look good, maybe a transparent floating window behind that of the nodes)
3. Show the jump-order through the snippet in a similar way, differently colored arrows would be very cool I think

One final, and more boring, problem would be storing the snippet: the most straightforward way is just writing them to a text-file and loading that, but it should be possible to store it as lua bytecode, which can just be loaded and appended to, than stored again.

**Comment by ziontee113** [comment]
I do plan to store the snippet as a regular .lua file. I plan to start small & simple, only supporting insert nodes & text-only choice nodes.

Speaking of Hydra, how are you using it? I'm trying it out to trigger LuaSnippets, but the big picture is still hazy atm.

**Reply by L3MON4D3** [reply]
> I do plan to store the snippet as a regular .lua file. I plan to start small & simple, only supporting insert nodes & text-only choice nodes.

 That sounds like a good approach👍
(And is probably a good bit of work already :D) 

> Speaking of Hydra, how are you using it? I'm trying it out to trigger LuaSnippets, but the big picture is still hazy atm.

Actually not at all😅
I wanted to use it to for debugging, but for me it doesn't add much, just takes up screen space.
What's your work flow for using it with snippets? To me, it seems like regular triggers for expansion+cmp for searching covers all one would want, but I'd love to see an improvement to that

**Reply by ziontee113** [reply]
Well, I'm finding a "better" way to trigger snippets / input text in general. I don't know if it's possible but I'm looking for a solution.

I'm experimenting a solution that has special movements & trigger snippets. Like a "sub-mode". That's where Hydra comes in. That plugin has so much potential and people (including me) can't exploit it to it's full potential, which makes me kinda mad :smile: 

I'll inform you if the result is any good at all. Things are still hazy atm.

**Reply by L3MON4D3** [reply]
Oh yeah, please do👍👍

**Reply by ziontee113** [reply]
Hi man @L3MON4D3 , I have a sample demo, it still feels so shit when there's 7 if statements on the screen :cry: 

https://user-images.githubusercontent.com/102876811/178470562-9aa4ab4b-0617-4029-bb1e-c8d1fd09df9f.mp4

Please tell me what you think. Is there any way I can improve this. It's driving me insane.

Also remember to turn audio on :sound:

**Reply by L3MON4D3** [reply]
Holy shit that wallpaper :laughing::laughing: Wow that's awesome, did not expect that
Omg just turned on sound I can't xD (did you add the sounds in post? if not, that's amazing, maybe do a plugin that hooks into some `vim.api`-functions to add some soundeffects, maybe `tts` on `nvim_buf_set_text` xD?)
Top Notch Demo :ok_hand::ok_hand:

For motion, do you know about [`hop`](https://github.com/phaazon/hop.nvim), or other easymotion-like plugins (sneak, lightspeed for example)? Makes it much easier to jump to different locations, but you'd still need something else to quickly expand snippets

**Reply by leiserfg** [reply]
This kinda sounds like the `otf` snippets, my original plan was to translate them into `fmt` snippets and store them on files but, at that point, we didn't have the lua  snippets loader, maybe now it's more ergonomic to do it. If you like the idea we could just make [this](https://github.com/L3MON4D3/LuaSnip/blob/7d78278c2a935b8cd1b6b43065223e14490f3133/lua/luasnip/extras/otf.lua#L24) public.

**Reply by L3MON4D3** [reply]
True, not dissimilar.. mhmmm
I'm not sure though, is it worth it to create yet another (lsp,lua) "fully-featured" (as opposed to the otf-snippet-language, which is more about quickly(!!) creating snippets) snippet-language?
